### PR TITLE
nvdaHelperRemote: Use SendMessageCallback instead of SendMessage to execute most cross-thread calls.

### PR DESCRIPTION
**Pull request by jcsteh**
This is necessary for Firefox multi-process. When we use SendMessage to execute calls in the main thread, outgoing cross-process COM calls fail with RPC_E_CANTCALLOUT_ININPUTSYNCCALL; see Mozilla bug 1297549 comment 14. We avoid this by instead using PostMessage and waiting on an event.
Even though this is only relevant to Firefox multi-process right now, this applies to any case where we might call a COM proxy in a cross-thread call.
execInWindow now uses SendMessageCallback. Virtual buffer backends now use execInWindow to initialize/terminate the render thread instead of their own message/hook.

### CODE REVIEW

**Comment by jcsteh** [comment]
@MichaelDCurran, can you please review? Thanks!

**Comment by jcsteh** [comment]
@MichaelDCurran, I updated this to also do this for execInWindow. I've changed the title and body of the PR accordingly.

File: nvdaHelper/vbufBase/backend.cpp [hunk 1]
```diff
@@ -36,11 +34,13 @@ void VBufBackend_t::initialize() {
...
-	LOG_DEBUG(L"Registered hook, sending message...");
-	SendMessage((HWND)UlongToHandle(rootDocHandle),wmRenderThreadInitialize,(WPARAM)this,0);
-	LOG_DEBUG(L"Message sent, unregistering hook");
-	unregisterWindowsHook(WH_CALLWNDPROC,renderThread_callWndProcHook);
+	auto func = [&] {
+		LOG_DEBUG(L"Calling renderThread_initialize on backend at "<<this);
+		this->renderThread_initialize();
+	};
+	LOG_DEBUG(L"Calling execInWindow");
+	execInWindow((HWND)UlongToHandle(rootDocHandle),func);
```
Comment (michaelDCurran) [inline nvdaHelper/vbufBase/backend.cpp]: I'm a little unsure if this is completely safe as it is a call from one instance of the CRT to another, using a c++ class or specific feature. I really have no evidence that it won't work, but just something we should keep an eye on in various operating systems.  For instance, we avoided ever passing c++ strings for this reason. At worse

**Comment by jcsteh** [comment]
As discussed, I tested this with Firefox, Chrome and IE in Windows 10, 
as well as Firefox and IE in Windows XP, and did not encounter any 
crashes or other instability. Hopefully, this multiple CRT thing won't 
be an issue, but if it is, it should get picked up during incubation.

**Comment by Brian1Gaff** [comment]
Would this affect anything else using IE code such as Outlook Express etc?
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by jcsteh** [comment]
I don't anticipate that it will break anything, but yes, this new code does get used in anything that uses IE.

**Comment by jcsteh** [comment]
@MichaelDCurran, could you please review this again?

**Comment by michaelDCurran** [comment]
Fine with me.

**Pull request merged**